spinalcordtoolbox/reports/assets/_assets/js/main.js

```diff
@@ -215,19 +224,79 @@ function responseHandler(res) {
 }
 
 function set_download_yml_btn_state(marker) {
+  // Return early if marker is null
+  if (marker === null) return;
+
   let disabled = true;
   sct_data.forEach(item => {
       if (item.qc === marker) {
         disabled = false;
       }
   });
+
   if (containsNonLatinCodepoints(marker) === true) {
     // This converts e.g. '\u2718' -> '2718' with corresponding id='download_yaml_btn_2718'
     marker = marker.codePointAt(0).toString(16)
   }
-  document.getElementById("download_yaml_btn_".concat(marker)).disabled = disabled;
+
+  // Safeguard accessing the DOM element
+  let buttonElement = document.getElementById("download_yaml_btn_".concat(marker));
+  if (buttonElement) {
+    buttonElement.disabled = disabled;
+  }
 }
 
 function containsNonLatinCodepoints(s) {
     return /[^\u0000-\u00ff]/.test(s);
 }
+
+function downloadQcStates() {
+  var qcFlags = {};
+  // Fetch all qc flags from local storage
+  sct_data.forEach(function(item, index) {
+      var uniqueId = item.moddate + '_' + item.fname_in + '_' + item.command;
+      var state = localStorage.getItem(uniqueId);
+      if (state) {
+          qcFlags[uniqueId] = state;
+      }
+  });
+  // Create a blob and trigger a download
+  var dataStr = "data:text/json;charset=utf-8," + encodeURIComponent(JSON.stringify(qcFlags));
+  var downloadAnchorNode = document.createElement('a');
+  downloadAnchorNode.setAttribute("href", dataStr);
+  downloadAnchorNode.setAttribute("download", "qc_flags.json");
+  document.body.appendChild(downloadAnchorNode);
+  downloadAnchorNode.click();
+  downloadAnchorNode.remove();
```

Comment:
As you mentioned, saving and loading a `.json` file isn't ideal, since it's a manual process that's separate from the QC folder, and thus won't be automatically zipped up when someone sends the entire QC folder to a colleague.

- However, maybe we can improve the user experience a little? What if, instead of saving and loading the `.json` file to the default location (e.g. `Downloads/`), we saved it to the QC folder?
- Then, since there is a hardcoded location where this json file should exist, we could check for the existence of this file on page load and load it automatically?
- This would be more "automatic". But, now that I think about it, your current solution has one big benefit: It allows for multiple different QC states to be stored on-disk (e.g. for different raters working on the same QC).
- So, maybe a simpler, easier fix could be to just start the file dialog in the QC folder (for both saving and loading), rather than defaulting to the `Downloads/` folder. That way, people naturally think to preserve the QC folder as the download location.

I'm not sure how to do this using JavaScript, though? In fact, I'm not even sure [if it's possible](https://stackoverflow.com/questions/11336663/how-to-make-a-browser-display-a-save-as-dialog-so-the-user-can-save-the-conten)? From what I can tell, there's a new-ish API for this ([`showSaveFilePicker()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/showSaveFilePicker)) but it's experimental and apparently not supported by Firefox.

Reply:
Ah! Apparently whether or not a "Save as..." dialog comes up is browser-specific. Dang. That's really inconvenient!